contracts/utils/StrikeSelection.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.3;
+pragma experimental ABIEncoderV2;
+
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+import {IOtoken} from "../interfaces/GammaInterface.sol";
+import {DSMath} from "../lib/DSMath.sol";
+import {IOptionsPremiumPricer} from "../interfaces/IRibbon.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract StrikeSelection is DSMath, Ownable {
+    using SafeMath for uint256;
+
+    /**
+     * Immutables
+     */
+    IOptionsPremiumPricer public immutable optionsPremiumPricer;
+
+    // delta for options strike price selection (ex: 0.1 is 1)
+    uint256 public delta;
+    // step pct at which we will iterate over
+    // (ex: 500 = 50% means we will move in leaps of 50% of spot price)
+    uint256 public step;
+
+    event DeltaSet(uint256 oldDelta, uint256 newDelta, address owner);
+    event StepSet(uint256 oldStep, uint256 newStep, address owner);
+
+    constructor(address _optionsPremiumPricer) {
+        require(_optionsPremiumPricer != address(0), "!_optionsPremiumPricer");
+        optionsPremiumPricer = IOptionsPremiumPricer(_optionsPremiumPricer);
+        // set delta to 0.1
+        delta = 1;
+        // set step to 5%
+        step = 50;
+    }
+
+    /**
+     * @notice Gets the strike price satisfying the delta value
+     * given the expiry timestamp and whether option is call or put
+     * @param expiryTimestamp is the unix timestamp of expiration
+     * @param isPut is whether option is put or call
+     */
+
+    function getStrikePrice(uint256 expiryTimestamp, bool isPut)
+        external
+        returns (uint256 strikePrice)
+    {
+        // asset price
+        uint256 assetPrice = optionsPremiumPricer.getUnderlyingPrice();
+
+        // time to expiration in number of days
+        uint256 t = expiryTimestamp.sub(block.timestamp).div(1 days);
+
+        // For each asset prices with step of 'margin' (down if put, up if call)
+        //   if that assets getOptionDelta(currStrikePrice, t) == (isPut ? 1 - delta : delta) with certain margin of error
+        //        return strike price
+
+        bool pastDelta = false;
+        uint256 currStrike = assetPrice;
+        uint256 newDelta = isPut ? uint256(10).sub(delta) : delta;
+
+        while (!pastDelta) {
```

Comment:
Love the approach on how to get strike price! A few comments:

- We should bound the number of iterations with a for loop instead of using a while loop. I haven't verified but it seems possible end up in an infinite loop
- Maybe fallback to whatever's the best strike price?

Reply:
So as it stands now, the delta will always be decreasing for calls and it will be increasing for puts. For calls we are looking for 0.1d and for puts 0.9d. When we iterate at each step we are either increasing or decreasing the delta. At some point we will reach the threshold of 0.1 and 0.9 and currDelta > newDelta will be true for puts or currDelta < newDelta will be true for calls.

We can add logic to just do % from spot price if while loop goes past x iterations? the only way this is possible is if chainlink asset oracle fails and latestAnswer() gives a wacky answer

Reply:
Thanks for the clarification, makes sense that the iteration is bounded even with a while loop because it's just monotonically increasing/decreasing